File: A2Z_striver/1 Basics/1.2 Patterns/pattern_17/pattern_17.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print the Alpha-Hill (Palindromic Char Pyramid).
// Example Input: n = 3
// Example Output:
//   A
//  ABA
// ABCBA
// Goal: O(N^2) Time.

class Solution
{
public:
    // ~ Approach 1: The Breakpoint Method (Standard)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Clean separation of spaces and character logic.
    void pattern17_breakpoint(int n)
    {
        for (int i = 0; i < n; i++)
        {

            // Spaces Logic: Standard (n - i - 1)
            for (int j = 0; j < n - i - 1; j++)
                cout << " ";

            // --- BREAKPOINT DERIVATION ---
            // 1. Total Width: The row follows odd numbers (1, 3, 5...) -> Formula: 2*i + 1
            // 2. The Turning Point:
            //    - We increment 'i' times.
            //    - We decrement 'i' times.
            //    - The Peak is the middle.
            //    - Math: (2*i + 1) / 2 simplifies to just 'i' in integer division.
            //      Example (i=2): Width 5. 5/2 = 2. We grow for 2 steps.

            int breakpoint = i;
            char ch = 'A';

            // Loop j from 1 to Total Width
            for (int j = 0; j < 2 * i + 1; j++)
            {
                cout << ch;

                // 3. CONDITION DERIVATION
                //    - If j <= breakpoint (e.g., j <= 2), we are still climbing the hill.
                //    - Else, we have passed the climbing phase and must slide down.
                if (j < breakpoint){
                    ch++;
                }else{
                    ch--;
                }
            }
            cout << "\n";
        }
    }

    // ~ Approach 2: Coordinate Logic (Your Way - Polished)
    // Analysis: Time O(N^2) | Space O(1)
    void pattern17_coordinate(int n)
    {
        for (int i = 0; i < n; i++)
        {
            char ch = 'A' - 1;

            // Peak is always at index (n-1).
            // Right Boundary = Peak + i
            // Formula: n - 1 + i
            for (int j = 0; j <= n - 1 + i; j++)
            {

                // 1. LEFT BOUNDARY CHECK (Spaces)
                // The pyramid starts 'i' steps left of Peak.
                // Start = Peak - i = (n - 1) - i
                // If j is less than Start, it's a space.
                if (j < n - 1 - i)
                {
                    cout << " ";
                }

                // 2. RISING SLOPE (Left of or at Peak)
                // If j is valid (>= Start) but <= Peak (n-1).
                else if (j <= n - 1)
                {
                    ch++;
                    cout << ch;
                }

                // 3. FALLING SLOPE (Right of Peak)
                // If j > Peak, we simply mirror the logic (decrement).
                else
                {
                    ch--;
                    cout << ch;
                }
            }
            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 4;

    cout << "--- Approach 1: Breakpoint (Standard) ---\n";
    sol.pattern17_breakpoint(n);

    cout << "\n--- Approach 2: Coordinate (Your Logic) ---\n";
    sol.pattern17_coordinate(n);

    return 0;
}

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_17/pattern_17.py
import sys


class Solution:
    """
    Problem: Print the Alpha-Hill (Palindromic Char Pyramid).
    Example Input: n = 3
    Example Output:
      A
     ABA
    ABCBA
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Breakpoint/Midpoint Logic (Your Solution)
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Intuitive. We explicitly count up, hit the middle, then count down.
    def pattern17_breakpoint(self, n):
        for i in range(n):
            # 1. Spaces
            for j in range(n - i - 1):
                print(" ", end="")
            
            # 2. Characters
            mid = i  # Peak is at index 'i' (e.g., if width is 3, indices are 0,1,2. Peak is 1)
            val = ord('A')
            
            for j in range(2 * i + 1):
                print(chr(val), end="")
                # If strictly left of mid -> Climb Up. Else -> Slide Down.
                val = val + 1 if (j < mid) else val - 1
            
            print()

    # // ~ Approach 2: Coordinate Geometry (Single Loop Math)
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Single inner loop handles spaces and chars using index boundaries.
    def pattern17_coordinate(self, n):
        for i in range(n):
            val = ord('A') - 1
            
            # // Iterate until the theoretical end of the pyramid for this row
            # // Boundary: Center(n-1) + offset(i)
            for j in range(n - 1 + i + 1):
                
                # // 1. LEADING SPACES
                # // Formula: The pyramid starts 'i' steps before the center (n-1).
                if j < n - 1 - i:
                    print(" ", end="")
                
                # // 2. RISING EDGE (Climbing Up)
                # // We are inside the pyramid, but <= Center (n-1).
                elif j <= n - 1:
                    val += 1
                    print(chr(val), end="")
                
                # // 3. FALLING EDGE (Sliding Down)
                # // We are past the Center.
                else:
                    val -= 1
                    print(chr(val), end="")
            
            print()

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    n = 3
    
    print(f"--- Approach 1: Breakpoint (n={n}) ---")
    solver.pattern17_breakpoint(n)
    
    print(f"\n--- Approach 2: Coordinate (n={n}) ---")
    solver.pattern17_coordinate(n)